evm/vm/base.py

```diff
@@ -255,19 +259,25 @@ def snapshot(self):
         """
         Perform a full snapshot of the current state of the VM.
 
-        TODO: This needs to do more than just snapshot the state_db but this is a start.
+        Snapshots are a combination of the state_root at the time of the
+        snapshot and the indices of the journaled DB.
         """
-        with self.state_db(read_only=True) as state_db:
-            return state_db.snapshot()
+        return (self.block.header.state_root, self.journal_db.snapshot())
 
     def revert(self, snapshot):
         """
-        Revert the VM to the state
-
-        TODO: This needs to do more than just snapshot the state_db but this is a start.
+        Revert the VM to the state at the snapshot
         """
+        state_root, journal_index = snapshot
+
         with self.state_db() as state_db:
-            return state_db.revert(snapshot)
+            # first revert the database state root.
+            state_db.root_hash = state_root
+            # now roll the underlying database back
+            self.journal_db.revert(journal_index)
+
+    def commit(self):
```

Comment:
This doesn't seem to be used anywhere.

Also, this is the only place where clear_journal() is called, so unless I'm missing something the journal will never be cleared?

Reply:
Good catch.